### DISCUSSION TITLE
Run code on MS Inventor Hub without PC

### DISCUSSION THREAD

**Discussion by lobodpav** [pybricks/support#573] [Q&A] [answered]
Is it possible to upload a code to the Mindstorms Inventor Hub and activate it without a PC being connected via BLE?

I.e. something like the original FW allows - upload a program under a number and then execute them by pressing the middle button.

**Comment by laurensvalk** [comment, marked as answer]
This is under development, see https://github.com/pybricks/pybricks-micropython/pull/75

**Reply by lobodpav** [reply]
Thank you! Is the code "stable enough" to try it out? Or shall I rather wait? 😊

**Reply by laurensvalk** [reply]
That part is stable enough, but there are some other things to keep in mind.

We recently removed the concept of dual-booting the hub and the online installer is not yet updated, so you'd need to use `pybricksdev` to try this. And since battery charging is not yet implemented, you'd have to switch back to recharge the battery, which isn't very convenient.

**Reply by lobodpav** [reply]
I see. Without battery charging that makes no sense to use now. Will wait then, thanks!

**Reply by lobodpav** [reply]
Wondering if I can help in any way with the charging issue? Since I work on the volume control on the `main` branch, I lost the dual-boot. This is very annoying since my Hub has been discharged and now, if I understand it right, I have to
1. Restore original Lego FW (found 1.2 on the web)
2. Charge the hub
3. Re-install the Pybricks FW

The dual-boot was quite convenient for the charging, but since it's gone, the charging issue deserves to get fixed as soon as possible.

Also, installing Pybricks for 1.4 Lego Inventor FW is not supported yet, so I can' even check battery percentage as the official Lego app gets blocked until I update to the latest 1.4 FW 😞 Is there a way to find out battery charge level per cent? There is a battery voltage API, but since I don't know min/max battery voltage, it is difficult to estimate it.

**Comment by dlech** [comment]
You can find the LEGO v1.4 firmware at <https://github.com/gpdaniels/spike-prime> and use `pybricksdev dfu restore xxxxx.bin` to flash it to the hub. 

The voltage levels are in https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/sys/battery.c.

**Reply by lobodpav** [reply]
Lovely, thanks!

**Reply by lobodpav** [reply]
I've tested that now.
1. The Hub reports `7262 mV` using `InventorHub().battery.voltage()`.
2. Original Lego Mindstorms app says the battery charge is `60%`.
3. With Pybricks firmware, the light gives `2x orange, 1x blue` flashes when turned off and connected via USB (which indicates voltage below `6800 mV`).

Something does not play well. Or maybe I've misunderstood something. The `C` code looks good though.

**Reply by laurensvalk** [reply]
> With Pybricks firmware, the light gives 2x orange, 1x blue flashes when turned off and connected via USB (which indicates voltage below 6800 mV).

This is currently expected because 5V < 6.8V. The light indicators will be addresses when we implement a battery charging driver with a corresponding UI (i.e. status light, etc).

**Reply by lobodpav** [reply]
> This is currently expected because 5V < 6.8V.

Not sure I'm following. The voltage was `7.2V` which is above `6.8V`.

**Reply by laurensvalk** [reply]
The hub is turned off but it is still powered by USB (5V). Either way, this will get addressed when the charging driver is implemented.

**Reply by lobodpav** [reply]
Ah, now I get it. Makes sense, thanks!

**Reply by lobodpav** [reply]
May I help with the charging driver? Is it rocket science or would I be able to code it?

**Reply by laurensvalk** [reply]
I think @dlech is currently working on it. It was paused due to an issue with vbus detection, but that seems to be resolved.

We'll certainly appreciate an extra pair of eyes to review it when it is ready, just to make sure everything is safe. Almost all safety aspects seem to be handled in hardware, but double checking shouldn't hurt.

**Reply by lobodpav** [reply]
I see. Thanks for clarifying. Will gladly help any time possible.

**Reply by lobodpav** [reply]
It would be for the voltage thresholds to be available in the `PrimeHub` as properties. Or maybe even provide a new `PrimeHub().battery.percentage()` API, what do you think?

**Reply by laurensvalk** [reply]
Hardcoded constants can be documented, but a `percentage` method is certainly nice. We have considered it several times, but haven't settled on a good definition beyond a simple voltage scale yet. Particularly for the other hubs where everyone uses different batteries.

**Reply by lobodpav** [reply]
The `percentage` API would make sense for the LI_ION batteries anyway. For hubs with AA batteries, the percentage can't be really calculated. That's why I was thinking of the `percentage` API for PrimeHub only (not sure if any other HUBS use LI_IONs).

Yeah, the voltage scale is where most end up anyway. But gives at least some estimate.
